### ISSUE TITLE
twisted.web2.test.test_xmlrpc should not use twisted.web.xmlrpc.Proxy

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1929]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#1929|
|Type|defect|
|Created|2006-07-16 18:43:17Z|

These tests are essentially invalid and should be rewritten:
 1. They use the network and don't disconnect properly ([#1906](https://github.com/twisted/twisted/issues/1906))
 1. Both the client and the server use xmlrpclib so if it's broken they can still pass

Attachments:

* [1929.xmlrpc.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d20/d20c6623e24cf210d24b32b8c04191c4013e18f0/0697f8524fbe278682fec7035ad62dd722a91aa5.patch) (2550 bytes) - added by oubiwann on 2006-12-02 15:37:30Z - 
* [1929.test_xmlrpc.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d20/d20c6623e24cf210d24b32b8c04191c4013e18f0/35766324eb65197460a323895434150eafe44a89.patch) (10155 bytes) - added by oubiwann on 2006-12-03 19:12:51Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1929 1929
type__defect defect
reporter__dreid dreid
priority__high high
milestone__Web2_Gold_Master Web2-Gold-Master
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1153075397000000 1153075397000000
changetime__1165383339000000 1165383339000000
version__None None
owner__ 
cc__oubiwann cc__dreid cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_3"></a>|@oubiwann set status to `assigned`|
|-|-|

More explanation from dreid on IRC tonight:
```
[9:12pm] dreid: your tests are still all over the network :(
[9:13pm] oubiwann: yeah, 1) I'm not even looking at netw vs. nonnetw tests
         yet (just focusing on proper disconnects)
[9:13pm] oubiwann: and 2) I'm doing that work in a separate branch
[9:13pm] dreid: well if you don't hit the network you don't have to worry 
         about disconnects.
[9:14pm] oubiwann: I'm not convinced that not hitting the network is a good 
         thing...
[9:14pm] oubiwann: but I'm willing to be
[9:23pm] dreid: :\
[9:24pm] dreid: for one thing, all the tests for the client include talking 
         to the server, and all your tests for the server inclue talking to 
         the client.
[9:25pm] dreid: you're not really proving that either one is correct.
[9:25pm] oubiwann: ah, I see what you were saying in the tracker item now...
```

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_5"></a>|@oubiwann commented|
|-|-|

I'm about half-way through rewriting the tests. The parts that are rewritten do not use the network, though they are still using ```xmlrpclib``` to do the heavy lifting of generating payload XML.

dreid and I discussed this for a while on IRC last night (the fact that ```xmlrpclib``` is being trusted to work). One suggestion I offered is that, in the event that it is important to prove the trustworthiness of ```xmlrpclib```, we write an additional unittest at the top of ```t.w2.xmlrpc``` specifically for testing ```xmlrpclib```'s XML-generation and parsing capabilities. With that done, we could then feel free to use ```xmlrpclib``` in the rest of the tests instead of building all the XML by hand.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_6"></a>|@oubiwann commented|
|-|-|

Converted the first introspection unit test (the last test class that needs converting); 2 more introspection tests to go (and then docstrings). Hopefully will finish this tonight.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_7"></a>|@oubiwann set owner to @dreid<br>@oubiwann set status to `new`|
|-|-|

David, patches have been uploaded (I don't have branch commit access). They are ready for your comments/suggestions/etc.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @oubiwann|
|-|-|

Some simple initial comments:

  * test_xmlrpc.py now imports t.web2.channel.http _and_ t.web2.client.http.  Some other facts pointed out by pyflakes:
```
twisted/web2/test/test_xmlrpc.py:12: 'server' imported but unused
twisted/web2/test/test_xmlrpc.py:14: redefinition of unused 'http' from line 13
twisted/web2/test/test_xmlrpc.py:14: 'http' imported but unused
twisted/web2/test/test_xmlrpc.py:17: 'unittest' imported but unused
twisted/web2/test/test_xmlrpc.py:18: 'reactor' imported but unused
twisted/web2/test/test_xmlrpc.py:21: 'http_headers' imported but unused
twisted/web2/test/test_xmlrpc.py:22: 'ClientTests' imported but unused
twisted/web2/test/test_xmlrpc.py:22: 'TestServer' imported but unused
```
   These should all be cleaned up.
  * There are some relative imports which should be changed to absolute imports.  Relative imports aren't allowed in Twisted.
  * \-style line continuations are also fairly undesirable.  These should use triple-quoted strings or unbalanced parenthesis instead.
  * There's a debug print in a nested function which isn't even used
  * Test methods named like test_MethodHelp should be test_methodHelp instead.
  * wait_timneout violates the naming convention and also seems to be completely ignored.  What's the point of this?
  *

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_9"></a>|@oubiwann set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_8):
> Some simple initial comments:
>
>   * Some other facts pointed out by pyflakes:
[snip]

Pyflakes now reports no badness for test/test_xmlrpc.py

>   * There are some relative imports which should be changed to absolute imports.  Relative imports aren't allowed in Twisted.

Done.

>   * \-style line continuations are also fairly undesirable.  These should use triple-quoted strings or unbalanced parenthesis instead.

Fixed (used parens).

>   * There's a debug print in a nested function which isn't even used

Both removed.

>   * Test methods named like test_MethodHelp should be test_methodHelp instead.

Changed.

>   * wait_timneout violates the naming convention and also seems to be completely ignored.  What's the point of this?

Dunno. I copied this from ```twisted.web2.test.test_server.BaseCase```. Removed in ```test_xmlrpc.py```.

Additionally, I ran Pyflakes against ```xmlrpc.py```, found some unused imports and removed those as well.

I've uploaded the newest changes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_11"></a>|@dreid set owner to @oubiwann|
|-|-|

* test_RPCFailure uses log.flushErrors which is deprecated. Sould use self.flushLoggedErrors instead.

* Extraneous whitespace in the list comprehensions in xmlrpc_listMethods

* Since TestServer and ClientTests really aren't used, there is no reason for their import to be there.  You can just remove the line completely.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_13"></a>|@oubiwann commented|
|-|-|

Replying to [dreid](#note_11):
> * test_RPCFailure uses log.flushErrors which is deprecated. Sould use self.flushLoggedErrors instead.
>
> * Extraneous whitespace in the list comprehensions in xmlrpc_listMethods
>
> * Since TestServer and ClientTests really aren't used, there is no reason for their import to be there.  You can just remove the line completely.

 * I'm assuming you meant ```self.flushErrors()```... I could only find one example of that in all of the Twisted source code, and that was in the tests for ```t.w2.test.test_wsgi```. I used that as an example.

 * Whitespace in list comprehensions fixed.

 * Commented import removed.

I will upload the new patches and set to review again.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann set owner to @dreid<br>@oubiwann set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_15"></a>|@dreid set owner to @oubiwann|
|-|-|

Sorry with self.flushLoggedErrors I was refering to twisted.trial.unittest.TestCase.flushLoggedErrors.  You should probably update you're working copy if you're not seeing the deprecation warning that explains this.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_16"></a>|@oubiwann commented|
|-|-|

Ah, got it -- thanks! Yeah, branch 1929 is enough out of date that the tests won't run with the newest trunk (fails on ```from twisted.application.reactors import Reactor```). I created my own local "branch" from the latest trunk, and now I see the deprecation message.

I fixed this to use ```TestCase.flushLoggedErrors``` and will shortly update this ticket with the latest patch.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_19"></a>|@dreid set status to `assigned`|
|-|-|

branched to web2-xmlrpc-tests-1929-2

builds on:
 * debian-py2.4-select
 * debian64-py2.4-select
 * osx-py2.4-select
 * osx-py2.4-select-gc
 * win32-py2.4-select 
 * win32-py2.5-select

unrelated failures on: 
 * win32-py2.4-er
 * debian-py2.5-select
 * suse-py2.5-select

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_20"></a>|@dreid set status to `closed`|
|-|-|

(In [18820]) Merge web2-xmlrpc-tests-1929-2

Author: oubiwann
Reviewer: dreid,exarkun
Fixes [#1929](https://github.com/twisted/twisted/issues/1929)

The Web2 XML-RPC tests no longer require twisted.web.xmlrpc.Proxy to run.  In fact they do not 
hit the network at all and so do not cause any Unclean Reactor warnings.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
